src/state.rs

```diff
@@ -672,6 +673,47 @@ impl DemandProxyState {
         fetch(addr)
     }
 
+    // same as fetch_workload, but if the caller knows the workload is enroute already,
+    // will retry on cache miss for a configured amount of time - returning the workload
+    // when we get it, or nothing if the timeout is exceeded, whichever happens first
+    pub async fn wait_for_workload(
+        &self,
+        addr: &NetworkAddress,
+        deadline: Duration,
+    ) -> Option<Arc<Workload>> {
+        debug!(%addr, "wait for workload");
+
+        // Take a watch listener *before* checking state (so we don't miss anything)
+        let mut wl_sub = self.state.read().unwrap().workloads.new_subscriber();
+
+        debug!(%addr, "got sub, waiting for workload");
+
+        if let Some(wl) = self.fetch_workload(addr).await {
+            return Some(wl);
+        }
+
+        // We didn't find the workload we expected, so
+        // loop until the subscriber wakes us on new workload,
+        // or we hit the deadline timeout and give up
+        let timeout = tokio::time::sleep(deadline);
+        let subscriber = wl_sub.changed();
```

Comment:
I think we could do something a bit more precise by re-using the on-demand infrastructure without actually using on-demand. Which basically amounts to using the `pending` map in XDS client to register a watch for a specific resource rather than anything.

I don't mind this either though

Reply:
Yep, could do per-resource watches easily enough. If this were (or becomes) more of a hotpath then we can do that.